### PR TITLE
Implement timeShift natively

### PR DESCRIPTION
**Pull request by shanson7**
**Describe your changes**
Implement timeShift natively

**Testing performed**
test files added
A/B testing against graphite using https://github.com/shanson7/mtcmptest

**Additional context**
There are a few scenarios that are a little strange:

1. `resetToEnd` does not seem to work when data is retrieved from MT. I think it's due to padding the data to the requested end.
2. `alignToDST` is hard to test on UTC servers :) It uses the server timezone (which I think is misleading to users) and that may be different in the Mt deploy than the graphite deploy

### CODE REVIEW

File: expr/validator.go [hunk 1]
```diff
@@ -37,6 +37,17 @@ func IsIntervalString(e *expr) error {
 	return err
 }
 
+func IsSignedIntervalString(e *expr) error {
+	durStr := e.str
+	if durStr[0] == '-' {
```
Comment (fqiu7) [inline expr/validator.go]: why not use
```
if (durStr[0] == '-' || durStr[0] == '+') {
    durStr = durStr[1:]
}
```

File: expr/func_timeshift.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+		signedOffset += dstOffset
+	}
+
+	return signedOffset, nil
+}
+
+func addOffset(orig uint32, offset int) uint32 {
+	if offset < 0 {
+		orig -= uint32(offset * -1)
```
Comment (fqiu7) [inline expr/func_timeshift.go:139]: so what happens if this result is negative (orig will be an incorrect positive value here)?

File: expr/func_timeshift.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	return context
+}
+
+func (s *FuncTimeShift) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	negativeOffset := -s.shiftOffset
```
Comment (fqiu7) [inline expr/func_timeshift.go:59]: is s.shiftOffset a signed offset?

File: expr/pool.go [hunk 4]
```diff
@@ -12,3 +16,13 @@ var pointSlicePool *sync.Pool
 func Pool(p *sync.Pool) {
 	pointSlicePool = p
 }
+
+// GetPooledSliceAtLeastSize returns a 0-len slice that can hold at least minSize Points
+func GetPooledSliceAtLeastSize(minSize int) []schema.Point {
+	out := pointSlicePool.Get().([]schema.Point)
+	if cap(out) < minSize {
+		// TODO - return `out` to pool? Could just fetch it again next iteration
+		out = make([]schema.Point, 0, minSize)
```
Comment (fqiu7) [inline expr/pool.go]: so in this case, out is not fetched out of the pool, it is more like calling make() locally

File: expr/validator.go (hunk 1 above)
Comment (shanson7) [reply expr/validator.go]: Good catch, this was copy/pasta from other code that had more handling.

File: expr/func_timeshift.go (hunk 2 above)
Comment (shanson7) [reply expr/func_timeshift.go:139]: Yes see comment at https://github.com/bloomberg/metrictank/pull/89/files/01ea8dd74d455203529e2456290e579b6cb12dc0#diff-a08ba41f2effbe7b76d8c3a9f891cd75R46

File: expr/pool.go (hunk 4 above)
Comment (shanson7) [reply expr/pool.go]: yes

File: expr/func_timeshift.go (hunk 3 above)
Comment (shanson7) [reply expr/func_timeshift.go:59]: yes, it is an int

File: expr/func_timeshift_test.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+		}
+		if len(g.Datapoints) != len(exp.Datapoints) {
+			t.Fatalf("case %q len output expected %d, got %d", name, len(exp.Datapoints), len(g.Datapoints))
+		}
+		for j, p := range g.Datapoints {
+			bothNaN := math.IsNaN(p.Val) && math.IsNaN(exp.Datapoints[j].Val)
+			if (bothNaN || p.Val == exp.Datapoints[j].Val) && p.Ts == exp.Datapoints[j].Ts {
+				continue
+			}
+			t.Fatalf("case %q: output point %d - expected %v got %v", name, j, exp.Datapoints[j], p)
```
Comment (agao48) [inline expr/func_timeshift_test.go]: no way to use `equalOutput`?

**Review by agao48 — APPROVED** [review]

**Review by fqiu7 — APPROVED** [review]
LGTM

**Comment by shanson7** [comment]
Merged in https://github.com/grafana/metrictank/pull/1873

**Pull request closed without merging**